### ISSUE TITLE
Attempting to send directory with broken symlinks causes error

### ISSUE THREAD

**Issue by maxalbert** [warner/magic-wormhole#112]
Attempting to send a directory that contains broken symlinks makes `wormhole` choke.

Steps to reproduce (tested under Python 3.5.2 on Mac OS X 10.11.6):

```
$ python --version
Python 3.5.2 :: Continuum Analytics, Inc.

$ wormhole --version
magic-wormhole 0.8.2+27.gd81f2da

$ mkdir some_dir
$ ln -s nonexisting_file.txt some_dir/broken_symlink.txt
$ wormhole send some_dir
```

Resulting error message:

```python
Building zipfile..
Traceback (most recent call last):
  File "/Users/maxalbert/miniconda3/envs/wormhole/lib/python3.5/site-packages/twisted/internet/defer.py", line 651, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/Users/maxalbert/miniconda3/envs/wormhole/lib/python3.5/site-packages/twisted/internet/defer.py", line 1355, in gotResult
    _inlineCallbacks(r, g, deferred)
  File "/Users/maxalbert/miniconda3/envs/wormhole/lib/python3.5/site-packages/twisted/internet/defer.py", line 1297, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/Users/maxalbert/miniconda3/envs/wormhole/lib/python3.5/site-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
--- <exception caught here> ---
  File "/Users/maxalbert/work/code/other/magic-wormhole/src/wormhole/cli/cli.py", line 103, in _dispatch_command
    yield maybeDeferred(command)
  File "/Users/maxalbert/miniconda3/envs/wormhole/lib/python3.5/site-packages/twisted/internet/defer.py", line 1297, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/Users/maxalbert/miniconda3/envs/wormhole/lib/python3.5/site-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/Users/maxalbert/code/magic-wormhole/src/wormhole/cli/cmd_send.py", line 55, in go
    yield d
builtins.FileNotFoundError: [Errno 2] No such file or directory: '/private/tmp/some_dir/broken_symlink.txt'
ERROR: [Errno 2] No such file or directory: '/private/tmp/some_dir/broken_symlink.txt'
```

**Comment by warner** [comment]
Oh, interesting. What do you think it should do instead here? We could make it behave like `tar`, where (I think) it'd send the (maybe dangling, maybe not) symlink target filename to the far side (where it might point at some surprising file). Or we could pre-scan the source directory for these problems before sending anything. Or we could omit the bad entry and print a warning.

**Comment by maxalbert** [comment]
If both the source and target support symbolic links then I'd personally prefer the first option (i.e., send symlinks no matter if they're dangling or not) because it basically creates an exact mirror of the source directory on the target side and therefore results in the least surprise for the user.

However, I don't know how Windows deals with symlinks so there may be hidden problems when sending a directory e.g. from Linux to Windows. I guess in this case either of the other options is fine, too.

I suspect that there is no way for `wormhole` to detect the target operating system? So we probably need to provide a command line switch to let the user choose which option they want. Any ideas how to best handle this from the UI point of view?

**Comment by warner** [comment]
Yeah, I can't think of a good way to learn the target filesystem properties ahead of time.

I'm a bit worried about symlinks that point outside the source tree (e.g. `~/.ssh/id_rsa`, something secret/sensitive on the target machine). What would you think about throwing an error (on both sides) if it sees a symlink whose target to something that isn't being copied?

**Comment by maxalbert** [comment]
Good point about potential malicious exploits using dangling symlinks.

To be honest, I'm not too keen on throwing errors because it is effectively the same as the current behaviour (just with a more meaningful error message). However, it gets in the way of usability because you'd have to trace down all dangling symlinks and delete them before you could send a directory. Also, there may potentially be good reasons why you don't want to (or can't) alter the source directory.

Therefore I think I'd prefer warnings over errors (and avoiding to copy over dangling symlinks).

For convenience, there could even be a switch to allow receiving dangling symlinks (if you're sending files to yourself), but that's likely overkill.

**Comment by warner** [comment]
@toolness and I worked on this one at PyCon .. we decided to go with the flag to turn the errors into mere warnings. Sending symlinks at all feels too likely to be surprising for the receiver. This way we can promise that the output of `wormhole receive` is strictly one or more normal files, with no effects or references to anything above the filename printed during the receive. Hopefully that strikes the right balance between usable and secure :).
